### PR TITLE
feat: form dialog extension

### LINKED ISSUE #4145: Add form dialog extension to Composer
Add the extension to Composer integrated with the build
OK to have placeholder UI as other tasks will cover adding UI and logic

### LINKED ISSUE #4146: Add form building UI to form dialog extension

### PR DESCRIPTION
**Pull request by hatpick**
## Description
Form dialog extension UI
This PR adds the UI for creating form dialog schema. Once the project tree is ready, form dialog will be integrated into the client web app.

## Task Item

closes #4145 
closes #4146 

## Screenshots
(demo)
sc1_fd_ext1

ezgif com-gif-maker

### CODE REVIEW

File: Composer/packages/form-dialogs/src/atoms/appState.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    const required = schemaPropertyStores.filter((property) => property.required).map((property) => property.name);
+    const examples = schemaPropertyStores.reduce<Record<string, string[]>>((acc, property) => {
+      if (property.examples?.length) {
+        acc[property.name] = property.examples;
+      }
+      return acc;
+    }, <Record<string, string[]>>{});
+
+    if (required.length) {
```
Comment (cwhitten) [inline Composer/packages/form-dialogs/src/atoms/appState.ts:130]: nit: explicit comparison opposed to falsey evaluation

File: Composer/packages/form-dialogs/src/atoms/types.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+export type RefPropertyPayload = TypedPropertyPayload & {
+  ref: string;
+};
+
+export type TypedPropertyPayload = {
+  kind: FormDialogPropertyKind;
+  entities?: string[];
+};
+
+const builtInFormats = ['data-time', 'date', 'time', 'email', 'uri', 'iri'] as const;
```
Comment (cwhitten) [inline Composer/packages/form-dialogs/src/atoms/types.ts:17]: what's `data-time` ?

File: Composer/packages/form-dialogs/src/atoms/utils.ts [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+      examples,
+      array,
+      payload,
+    };
+  });
+
+  return { name: schemaName, properties: propertyStores };
+};
+
+const findFirstMissingIndex = (arr: number[], start: number, end: number): number => {
```
Comment (cwhitten) [inline Composer/packages/form-dialogs/src/atoms/utils.ts:98]: i may have to pull you into interviews if i continue to see these clean common algorithms :)

File: Composer/packages/form-dialogs/src/atoms/utils.ts [hunk 4]
```diff
@@ -0,0 +1,289 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (cwhitten) [inline Composer/packages/form-dialogs/src/atoms/utils.ts:1]: I see some stuff here that doesn't look specific to form dialogs. i'd like to:

1. document the util functions a bit more
2. have a conversation to understand if any of these should be pulled in from a higher scope

**Review by cwhitten — COMMENTED** [review]
submitting comments so far. this diff will take awhile to go through

File: Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import { useHandlers } from 'src/atoms/handlers';
+import { VisualEditor } from 'src/components/VisualEditor';
+
+export type FormDialogSchemaEditorProps = {
+  /**
+   * Unique id for the visual editor.
+   */
+  editorId: string;
+  /**
+   * Wether to show or hide the theme picker.
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx:17]: Nit:  Whether

File: Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  showThemePicker?: boolean;
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx:19]: What is a theme picker for this UI?

File: Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  }, [editorId]);
+
+  const startOver = React.useCallback(() => {
+    reset({ name: editorId });
+  }, [reset, editorId]);
+
+  useRecoilTransactionObserver_UNSTABLE(async ({ snapshot, previousSnapshot }) => {
+    const content = await snapshot.getPromise(formDialogSchemaJsonSelector);
+    const prevContent = await previousSnapshot.getPromise(formDialogSchemaJsonSelector);
+    if (content !== prevContent) {
```
[the remark this answers is not in the document]
Comment (GeoffCoxMSFT) [reply Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx:70]: The getPromise is getting a snapshot of just the value of that one selector value.  I think this is the schema JSON string, so it ends up comparing two JSON strings to see if that has changed.  It isn't comparing two snapshot classes.  The comparison is necessary because the transaction observer fires for all the other changes going on in the app under the same RecoilRoot and it would cause the onSchemaUpdated to fire constantly without limiting the change detection to just the schema.

File: Composer/packages/form-dialogs/src/atoms/types.ts (hunk 2 above)
Comment (GeoffCoxMSFT) [reply Composer/packages/form-dialogs/src/atoms/types.ts:17]: Type of date-time (based on builtInStringFormats below)

File: Composer/packages/form-dialogs/src/atoms/utils.ts (hunk 4 above)
Comment (GeoffCoxMSFT) [reply Composer/packages/form-dialogs/src/atoms/utils.ts:1]: 3. It would be nice to have some unit tests around the utility functions that non-trivial (are doing indexing math into arrays).

File: Composer/packages/form-dialogs/src/components/VisualEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/components/VisualEditor.tsx:1]: Consider calling this something other than VisualEditor as that is the name of the component for the trigger/action flow editing.  FormDialogEditor? FormBuilder?

File: Composer/packages/form-dialogs/src/components/VisualEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+  formDialogSchemaAtom,
+  formDialogSchemaJsonSelector,
+  formDialogSchemaValidSelector,
+} from 'src/atoms/appState';
+import { useHandlers } from 'src/atoms/handlers';
+import { CommandBarUploadButton } from 'src/components/common/CommandBarUpload';
+import { FormDialogSchemaDetails } from 'src/components/property/FormDialogSchemaDetails';
+import { useUndoKeyBinding } from 'src/utils/hooks/useUndoKeyBinding';
+
+const downloadFile = async (fileName: string, schemaExtension: string, content: string) => {
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/components/VisualEditor.tsx:24]: Should this be in utils?

File: Composer/packages/form-dialogs/src/components/common/CommandBarUpload.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/components/common/CommandBarUpload.tsx:1]: Should file name be CommandBarUploadButton ?

File: Composer/packages/form-dialogs/package.json [hunk 11]
```diff
@@ -0,0 +1,63 @@
+{
+  "name": "@bfc/form-dialogs",
+  "version": "1.0.0",
```
[the remark this answers is not in the document]
Comment (hatpick) [reply Composer/packages/form-dialogs/package.json:3]: Sure.

File: Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (hatpick) [reply Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx:70]: It's comparing two string value of the schema in json.

File: Composer/packages/form-dialogs/src/atoms/types.ts (hunk 2 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/atoms/types.ts:17]: typo indeed

File: Composer/packages/form-dialogs/src/components/property/PropertyTypeSelector.tsx [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    return processSelection(
+      false,
+      (payload as ArrayPropertyPayload).items.kind,
+      newKind,
+      selectedKey,
+      (payload as ArrayPropertyPayload).items
+    );
+  }
+
+  if (oldKind === newKind) {
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/components/property/PropertyTypeSelector.tsx:40]: I think a comment is needed here to understand what happens when oldKind === newKind. Most code that has this comparison is optimizing to a no-op when there isn't a change, but you return a different result than when there is a change.

File: Composer/packages/form-dialogs/src/atoms/utils.ts (hunk 3 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/atoms/utils.ts:98]: sounds good, Geoff knows ;)

File: Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+      onGenerateDialog={onGenerateDialog}
+      onReset={startOver}
+    />
+  );
+});
+
+export const FormDialogSchemaEditor = (props: FormDialogSchemaEditorProps) => {
+  return (
+    <RecoilRoot>
+      <InternalFormDialogSchemaEditor {...props}></InternalFormDialogSchemaEditor>
```
[the remark this answers is not in the document]
Comment (hatpick) [reply Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx:89]: Internal component is for perf reasons, they both use the same prop type, aka it's safe.

File: Composer/packages/form-dialogs/src/atoms/utils.ts (hunk 4 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/atoms/utils.ts:1]: Other than the function giving the first available index for clone naming, everything else is specific to form dialog json schema. we can definitely move the other function up if there are other usecases.

File: Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx (hunk 6 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx:19]: Good catch, I removed theming infra and forgot to remove this

File: Composer/packages/form-dialogs/src/demo/Demo.tsx [hunk 14]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/demo/Demo.tsx:1]: Confirming - do you want to check in the Demo to Composer?

File: Composer/packages/form-dialogs/src/components/VisualEditor.tsx (hunk 8 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/components/VisualEditor.tsx:1]: FormDialogPropertiesEditor?

File: Composer/packages/form-dialogs/src/components/VisualEditor.tsx (hunk 9 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/components/VisualEditor.tsx:24]: not being re-used, so for now no :D

File: Composer/packages/form-dialogs/src/demo/Demo.tsx (hunk 14 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/demo/Demo.tsx:1]: Yes, it makes future work on the extension much faster.

File: Composer/packages/form-dialogs/tools/devServer.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
+'use strict';
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/tools/devServer.js:1]: Should this be part of the demo/ ?

Comment (hatpick) [reply Composer/packages/form-dialogs/tools/devServer.js:1]: It's under tools

File: Composer/packages/server/src/locales/en-US.json [hunk 16]
```diff
@@ -404,6 +425,9 @@
   "chris_whitten_11df1f35": {
     "message": "Chris Whitten"
```
Comment (GeoffCoxMSFT) [inline Composer/packages/server/src/locales/en-US.json:426]: Not part of this PR, but fun to localize - K’ris Uit’en

File: Composer/packages/server/src/locales/en-US.json [hunk 17]
```diff
@@ -1310,6 +1355,12 @@
   "manifest_version_1edc004a": {
     "message": "Manifest Version"
   },
+  "maximum_f0e8e5e4": {
+    "message": "Maximum"
+  },
+  "maximum_help_text_7f3e6562": {
+    "message": "Maximum help text"
```
Comment (GeoffCoxMSFT) [inline Composer/packages/server/src/locales/en-US.json:1362]: This is an odd message.

Comment (hatpick) [reply Composer/packages/server/src/locales/en-US.json:1362]: These are placeholder message until UX comes up with one.

**Comment by GeoffCoxMSFT** [comment]
I think this PR won't affect the current UI as it isn't wired into Composer top-level, correct?  When we get to the PR does does make this show up, let's be sure it is behind the feature switching mechanism.

**Review by GeoffCoxMSFT — APPROVED** [review]

File: Composer/packages/server/src/locales/en-US.json (hunk 16 above)
Comment (hatpick) [reply Composer/packages/server/src/locales/en-US.json:426]: lol

**Review by GeoffCoxMSFT — APPROVED** [review]

**Comment by hatpick** [comment]
> I think this PR won't affect the current UI as it isn't wired into Composer top-level, correct? When we get to the PR does does make this show up, let's be sure it is behind the feature switching mechanism.

@cwhitten I'm going to merge this PR since as Geoff has said it too, it will not affect any part of the Composer. I'm open to further comments and suggestions and will PR them  separately if there were any.
